### DISCUSSION TITLE
Grain boundary projection tensors

### DISCUSSION THREAD

**Discussion by wenpeng231024** [idaholab/moose#18782] [Q&A General] [answered]
Dear MOOSE group,

I try to build the diffusivity of bulk, surface, and grain boundary following the Eqs. (12)-(18) in the paper Modelling Simul. Mater. Sci. Eng. 23 (2015) 045009.

https://iopscience.iop.org/article/10.1088/0965-0393/23/4/045009/meta

In MOOSE, how to calculate the GB projection tensors as following? Or is there a related example?

![image](https://user-images.githubusercontent.com/60105926/132375729-c43627cf-d8da-4572-8461-c171e128c352.png)

![image](https://user-images.githubusercontent.com/60105926/132375572-f3363e3e-580a-4e64-beb5-73766f35408e.png)

Thanks.

Peng

**Comment by GiudGiud** [comment]
@hugary1995 @dschwen

**Comment by hugary1995** [comment, marked as answer]
The doxygen documentation is very useful in this case. [https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html](https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html) 

For example, you may use the `vectorOuterProduct` method to compute `Ts`. You can construct a second order identity tensor using
`RankTwoTensor I(RankTwoTensor::initIdentity)`

**Reply by wenpeng231024** [reply]
Many thanks! I will try this.

**Comment by laagesen** [comment]
You should be able to use the material `PolycrystalDiffusivityTensorBase` in the phase-field module:

moose/modules/phase_field/src/materials/PolycrystalDiffusivityTensorBase.C
